lib/components/base_component.dart

```diff
@@ -0,0 +1,147 @@
+import 'dart:ui';
+
+import 'package:meta/meta.dart';
+import 'package:ordered_set/comparing.dart';
+import 'package:ordered_set/ordered_set.dart';
+
+import '../effects/effects.dart';
+import '../effects/effects_handler.dart';
+import '../extensions/vector2.dart';
+import '../game.dart';
+import '../text_config.dart';
+import 'component.dart';
+
+/// This can be extended to represent a basic Component for your game.
+///
+/// The difference between this and [Component] is that this can have children,
+/// handles effects on your component and can be used to see whether a point is
+/// on your component, which is useful for handling the effect of gestures.
+abstract class BaseComponent extends Component {
+  final EffectsHandler _effectsHandler = EffectsHandler();
+
+  final OrderedSet<Component> children =
+      OrderedSet(Comparing.on((c) => c.priority));
+
+  /// This is set by the BaseGame to tell this component to render additional debug information,
+  /// like borders, coordinates, etc.
+  /// This is very helpful while debugging. Set your BaseGame debugMode to true.
+  /// You can also manually override this for certain components in order to identify issues.
+  bool debugMode = false;
+
+  Color get debugColor => const Color(0xFFFF00FF);
+
+  Paint get debugPaint => Paint()
+    ..color = debugColor
+    ..strokeWidth = 1
+    ..style = PaintingStyle.stroke;
+
+  TextConfig get debugTextConfig => TextConfig(color: debugColor, fontSize: 12);
+
+  /// This method is called periodically by the game engine to request that your component updates itself.
+  ///
+  /// The time [dt] in seconds (with microseconds precision provided by Flutter) since the last update cycle.
+  /// This time can vary according to hardware capacity, so make sure to update your state considering this.
+  /// All components on [BaseGame] are always updated by the same amount. The time each one takes to update adds up to the next update cycle.
+  @mustCallSuper
+  @override
+  void update(double dt) {
+    _effectsHandler.update(dt);
+  }
+
+  @mustCallSuper
+  @override
+  void render(Canvas canvas) {
+    prepareCanvas(canvas);
+
+    if (debugMode) {
+      renderDebugMode(canvas);
+    }
+
+    canvas.save();
+    children.forEach((c) => _renderChild(canvas, c));
+    canvas.restore();
+  }
+
+  void renderDebugMode(Canvas canvas) {}
+
+  void _renderChild(Canvas canvas, Component c) {
+    if (!c.loaded) {
+      return;
+    }
+    c.render(canvas);
+    canvas.restore();
+    canvas.save();
+  }
+
+  @protected
+  void prepareCanvas(Canvas canvas) {}
```

Comment:
is this needed here? Isn't this only needed by the `PositionComponent`?

Reply:
If we place it in `PositionComponent` and `super.render` is called first, the `prepareCanvas` will be called too late. I think it could potentially be useful for other implementations too that have anchor-like things. But that is mostly theoretical, if you have a good way to move it to `PositionComponent` I'm all ears.

Reply:
Thanks for clarifying, makes sense for me now.